authentication/active_directory_authentication_provider.py

```diff
@@ -0,0 +1,53 @@
+#########
+# Copyright (c) 2016 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+
+from ldap_authentication_provider import LDAPAuthenticationProvider
+
+DEFAULT_AUTH_ATTR_NAME = 'sAMAccountName'
+
+
+class ActiveDirectoryAuthenticationProvider(LDAPAuthenticationProvider):
+
+    def __init__(self, args, domain_name=None, **kwargs):
+        if 'search_props' in kwargs:
+            if 'user_id_attribute' not in kwargs.get('search_props'):
+                kwargs['search_props']['user_id_attribute'] = \
+                    DEFAULT_AUTH_ATTR_NAME
+        super(ActiveDirectoryAuthenticationProvider, self).__init__(args,
+                                                                    **kwargs)
+        self.domain_name = domain_name
+
+    def authenticate(self, userstore=None):
+        who, password = self.get_creds_from_request()
+
+        if self.is_search_based_authentication():
```

Comment:
we authenticate by binding either way, it's just that we sometimes have to get the user object to retrieve the dn. so, we should ask:
    if self.is_search_based_authentication():
        who = search_and_get_dn(...)

and then just call validate_bind, regardless of search being done before it or not:
self.validate_bind(user_dn, password)